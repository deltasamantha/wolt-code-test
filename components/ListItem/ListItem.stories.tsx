import {Meta, StoryObj} from "@storybook/react-native";
import {View} from "react-native";

import ListItem, {Props} from "./ListItem";
import {storyStyles} from "./ListItem.styles";
import {DayType} from "../../models/OpeningHours";
import {Slot} from "../../models/TimeSlot";

const mockedSlots: Slot[] = [
  {
    open: 36000,
    close: 43200,
  },
  {
    open: 43500,
    close: 75600,
  },
];

const ListItemMeta: Meta<Props> = {
  title: "ListItem",
  component: ListItem,
  argTypes: {
    day: {
      name: "Day type",
      required: true,
      description: "Day type",
      control: {
        type: "radio",
      },
      options: {
        monday: "monday" as DayType,
        tuesday: "tuesday" as DayType,
        wednesday: "wednesday" as DayType,
        thursday: "thursday" as DayType,
        friday: "friday" as DayType,
        saturday: "saturday" as DayType,
        sunday: "sunday" as DayType,
      },
    },
    isToday: {
      description: "Is today",
      control: {
        type: "boolean",
      },
    },
    slots: {
      description: "Available slots for the day",
      control: "object",
    },
  },
  decorators: [
    Story => (
      <View style={storyStyles.rootContainer}>
        <Story />
      </View>
    ),
  ],
};

export default ListItemMeta;

export const Primary: StoryObj<Props> = {
  args: {
    day: "monday",
    slots: [],
  },
};

export const TodayItemWithNoSlots: StoryObj<Props> = {
  args: {
    ...Primary.args,
    isToday: true,
  },
};

export const TodayItemWithSingleSlot: StoryObj<Props> = {
  args: {
    ...Primary.args,
    isToday: true,
    slots: mockedSlots.slice(1),
  },
};

export const ItemWithMultipleSlot: StoryObj<Props> = {
  args: {
    ...Primary.args,
    isToday: true,
    slots: mockedSlots,
  },
};
