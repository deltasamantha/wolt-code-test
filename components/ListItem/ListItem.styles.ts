import {StyleSheet} from "react-native";

import theme from "../../theme/theme";

export const styles = StyleSheet.create({
  rootContainer: {
    backgroundColor: theme.colors.colorWhite,
    paddingVertical: theme.spacing.md,
    paddingHorizontal: theme.spacing.lg,
    flexDirection: "row",
    justifyContent: "space-between",
    alignItems: "flex-start",
  },
  dateContainer: {
    flexDirection: "row",
    alignItems: "center",
  },
  todayIndicator: {
    paddingStart: theme.spacing.md,
  },
});

export const storyStyles = StyleSheet.create({
  rootContainer: {
    flex: 1,
    paddingHorizontal: theme.spacing.lg,
    backgroundColor: theme.colors.colorBackground,
    justifyContent: "center",
  },
});
