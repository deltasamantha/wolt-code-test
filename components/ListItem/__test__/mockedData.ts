import {Slot} from "../../../models/DisplayTimeSlot";

export const slotsData: Slot[] = [
  {
    open: 36000,
    close: 43200,
  },
  {
    open: 43500,
    close: 75600,
  },
];
