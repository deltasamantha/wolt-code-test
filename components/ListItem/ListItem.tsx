import React from "react";
import {View} from "react-native";
import Animated, {ComplexAnimationBuilder} from "react-native-reanimated";

import {styles} from "./ListItem.styles";
import locale from "../../config/Locale";
import {DayType} from "../../models/OpeningHours";
import {Slot} from "../../models/TimeSlot";
import AppText from "../AppText/AppText";
import SlotsList from "../SlotsList/SlotsList";

/**
 * Prop definition for ListItem component
 */
export interface Props {
  /**
   * day to be shown in the item, one of
   * ```"monday" | "tuesday" | "wednesday" | "thursday" | "friday" | "saturday" | "sunday"```
   */
  day: DayType;
  /**
   * ```{open: number; close?: number;}[]```
   */
  slots: Slot[];
  isToday?: boolean;
  entering?: ComplexAnimationBuilder;
}

/**
 * ### ListItem component
 * -------------
 * @description Component to list down opening time slots against the dates
 */
const ListItem: React.FC<Props> = ({day, slots, isToday, entering}: Props) => {
  return (
    <Animated.View style={styles.rootContainer} entering={entering}>
      <View style={styles.dateContainer}>
        <AppText size="m" type="semiBold">
          {locale.weekDay[day]}
        </AppText>
        {isToday && (
          <AppText
            size="s"
            type="bold"
            color="accent"
            style={styles.todayIndicator}>
            {locale.today.toUpperCase()}
          </AppText>
        )}
      </View>
      <SlotsList slots={slots} />
    </Animated.View>
  );
};

export default ListItem;
