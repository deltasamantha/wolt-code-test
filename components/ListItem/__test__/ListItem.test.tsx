import {render, cleanup, screen} from "@testing-library/react-native";

import {slotsData} from "./mockedData";
import locale from "../../../config/Locale";
import ListItem from "../ListItem";

afterEach(cleanup);

describe("ListItem", () => {
  it("Should render ListItem properly", () => {
    const {toJSON} = render(<ListItem day="monday" slots={[]} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("ListItem should have a week day", () => {
    const mondayText = locale.weekDay.monday;
    render(<ListItem day="monday" slots={[]} />);
    const dayIndicatorText = screen.getByText(mondayText);
    expect(dayIndicatorText.props.children).toBeDefined();
    expect(dayIndicatorText.props.children).toEqual(mondayText);
  });

  it("ListItem should have a no slots indicator in case no slots available", () => {
    const closedText = locale.closed;
    render(<ListItem day="monday" slots={[]} />);
    const closedIndicatorText = screen.getByText(closedText);
    expect(closedIndicatorText).toBeDefined();
    expect(closedIndicatorText.props.children).toEqual(closedText);
  });

  it("ListItem should have today indicator for today", () => {
    const todayText = locale.today.toUpperCase();
    render(<ListItem day="monday" slots={[]} isToday />);
    const todayIndicatorText = screen.getByText(todayText);
    expect(todayIndicatorText.props.children).toBeDefined();
    expect(todayIndicatorText.props.children).toEqual(todayText);
  });

  it("ListItem should have slots if slots available", () => {
    render(<ListItem day="monday" slots={slotsData} isToday />);
    const slots = screen.getAllByText("PM", {exact: false});
    expect(slots).toBeTruthy();
    expect(slots.length).toBe(2);
  });
});
