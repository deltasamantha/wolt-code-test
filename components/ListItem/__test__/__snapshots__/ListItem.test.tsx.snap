// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem Should render ListItem properly 1`] = `
<View
  style={
    {
      "alignItems": "flex-start",
      "backgroundColor": "#FFFFFF",
      "flexDirection": "row",
      "justifyContent": "space-between",
      "paddingHorizontal": 16,
      "paddingVertical": 10,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#202125",
            "fontFamily": "Roboto_500Medium",
            "fontSize": 16,
            "lineHeight": 22,
          },
          undefined,
        ]
      }
    >
      Monday
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "flex-end",
        "flexDirection": "column",
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#A1A2A4",
            "fontFamily": "Roboto_400Regular",
            "fontSize": 16,
            "lineHeight": 22,
          },
          undefined,
        ]
      }
    >
      Closed
    </Text>
  </View>
</View>
`;
